File: camase/kalman.py
"""Integrated-random-walk Kalman filter with Joseph covariance update."""

from __future__ import annotations

from dataclasses import dataclass, field

import numpy as np


def transition(dt: float) -> np.ndarray:
    return np.array([[1.0, dt], [0.0, 1.0]], dtype=np.float64)


def process_cov(sigma_a2: float, dt: float) -> np.ndarray:
    dt2 = dt * dt
    dt3 = dt2 * dt
    return sigma_a2 * np.array(
        [[dt3 / 3.0, dt2 / 2.0], [dt2 / 2.0, dt]], dtype=np.float64
    )


H = np.array([[1.0, 0.0]], dtype=np.float64)
I2 = np.eye(2, dtype=np.float64)


@dataclass
class KalmanIRW:
    x: np.ndarray = field(default_factory=lambda: np.zeros(2, dtype=np.float64))
    P: np.ndarray = field(default_factory=lambda: np.diag([1e-4, 1e-6]).astype(np.float64))
    last_S: float = 1.0
    last_innov: float = 0.0
    last_nis: float = 1.0

    def copy(self) -> "KalmanIRW":
        k = KalmanIRW()
        k.x = self.x.copy()
        k.P = self.P.copy()
        k.last_S = self.last_S
        k.last_innov = self.last_innov
        k.last_nis = self.last_nis
        return k

    def predict(self, F: np.ndarray, Q: np.ndarray) -> None:
        self.x = F @ self.x
        self.P = F @ self.P @ F.T + Q
        self.P = 0.5 * (self.P + self.P.T)

    def update_joseph(self, y: float, R: float) -> float:
        innov = float(y - (H @ self.x)[0])
        S = float((H @ self.P @ H.T)[0, 0] + R)
        if S <= 1e-18:
            S = 1e-18
        K = (self.P @ H.T) / S
        self.x = self.x + K.ravel() * innov
        IKH = I2 - K @ H
        self.P = IKH @ self.P @ IKH.T + (K * R) @ K.T
        self.P = 0.5 * (self.P + self.P.T)
        self.last_S = S
        self.last_innov = innov
        self.last_nis = innov * innov / S
        return self.last_nis

    def one_step_pred(self, F: np.ndarray) -> float:
        return float((H @ (F @ self.x))[0])

File: src/lib/camase/kalman.ts
export type Vec2 = [number, number];

export function transition(dt: number): number[][] {
  return [
    [1, dt],
    [0, 1],
  ];
}

export function processCov(sa2: number, dt: number): number[][] {
  const dt2 = dt * dt;
  const dt3 = dt2 * dt;
  return [
    [(sa2 * dt3) / 3, (sa2 * dt2) / 2],
    [(sa2 * dt2) / 2, sa2 * dt],
  ];
}

function mm(A: number[][], B: number[][]): number[][] {
  return [
    [A[0][0] * B[0][0] + A[0][1] * B[1][0], A[0][0] * B[0][1] + A[0][1] * B[1][1]],
    [A[1][0] * B[0][0] + A[1][1] * B[1][0], A[1][0] * B[0][1] + A[1][1] * B[1][1]],
  ];
}

function mt(A: number[][]): number[][] {
  return [
    [A[0][0], A[1][0]],
    [A[0][1], A[1][1]],
  ];
}

function madd(A: number[][], B: number[][]): number[][] {
  return [
    [A[0][0] + B[0][0], A[0][1] + B[0][1]],
    [A[1][0] + B[1][0], A[1][1] + B[1][1]],
  ];
}

export class KalmanIRW {
  x: Vec2 = [0, 0];
  P: number[][] = [
    [1e-4, 0],
    [0, 1e-6],
  ];
  lastS = 1;
  lastInnov = 0;
  lastNis = 1;

  predict(F: number[][], Q: number[][]) {
    this.x = [F[0][0] * this.x[0] + F[0][1] * this.x[1], F[1][0] * this.x[0] + F[1][1] * this.x[1]];
    const FP = mm(F, this.P);
    this.P = madd(mm(FP, mt(F)), Q);
    const s = 0.5 * (this.P[0][1] + this.P[1][0]);
    this.P[0][1] = s;
    this.P[1][0] = s;
  }

  updateJoseph(y: number, R: number): number {
    const innov = y - this.x[0];
    let S = this.P[0][0] + R;
    if (S <= 1e-18) S = 1e-18;
    const K0 = this.P[0][0] / S;
    const K1 = this.P[1][0] / S;
    this.x[0] += K0 * innov;
    this.x[1] += K1 * innov;
    const IKH = [
      [1 - K0, 0],
      [-K1, 1],
    ];
    const mid = mm(mm(IKH, this.P), mt(IKH));
    const krk = [
      [K0 * R * K0, K0 * R * K1],
      [K1 * R * K0, K1 * R * K1],
    ];
    this.P = madd(mid, krk);
    const s = 0.5 * (this.P[0][1] + this.P[1][0]);
    this.P[0][1] = s;
    this.P[1][0] = s;
    this.lastS = S;
    this.lastInnov = innov;
    this.lastNis = (innov * innov) / S;
    return this.lastNis;
  }

  oneStepPred(F: number[][]): number {
    return F[0][0] * this.x[0] + F[0][1] * this.x[1];
  }
}
